text/0000-show-string-guidelines.md

```diff
@@ -0,0 +1,166 @@
+- Start Date: 2015-01-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+A [recent RFC](https://github.com/rust-lang/rfcs/pull/504) split what was
+previously `fmt::Show` into two traits, `fmt::Show` and `fmt::String`, with
+format specifiers `{:?}` and `{}` respectively.
+
+That RFC did not, however, establish complete conventions for when to implement
+which of the traits, nor what is expected from the output.  That's what this RFC
+seeks to do.
+
+It turns out that, due to the suggested conventions and other
+concerns, renaming the traits is also desirable.
+
+# Motivation
+
+Part of the reason for splitting up `Show` in the first place was some tension
+around the various use cases it was trying to cover, and the fact that it could
+not cover them all simultaneously. Now that the trait has been split, this RFC
+aims to provide clearer guidelines about their use.
+
+# Detailed design
+
+The design of the conventions stems from two basic desires:
+
+1. It should be easy to generate a debugging representation of
+   essentially any type.
+
+2. It should be possible to create user-facing text output via convenient
+   interpolation.
+
+Part of the premise behind (2) is that user-facing output cannot automatically
+be "composed" from smaller pieces of user-facing output (via, say,
+`#[derive]`). Most of the time when you're preparing text for a user
+consumption, the output needs to be quite tailored, and interpolation via
+`format` is a good tool for that job.
+
+As part of the conventions being laid out here, the RFC proposes to:
+
+1. Rename `fmt::Show` to `fmt::Debug`, and
+2. Rename `fmt::String` to `fmt::Display`.
+
+## Debugging: `fmt::Debug`
+
+The `fmt::Debug` trait is intended for debugging. It should:
+
+* Be implemented on every type, usually via `#[derive(Debug)]`.
+* Never panic.
```

Comment:
As @sfackler mentioned I believe it's important to reserve the use of `Error` to only write errors, not semantic errors. The code snippet you pointed out using `String` is "ok" because there's prior knowledge that `String` never returns an `Error` in its `write_str` implementation.

For an example usage of `Error`, see [`write_str` on `Writer`](https://github.com/rust-lang/rust/blob/master/src/libstd/io/mod.rs#L1016-L1041) for example.